Add support for event entities

kevdliu opened the pull request:
This PR creates an event entity for every button on a device. The event entities will be triggered when the corresponding button is triggered (pressed, doubleTapped, held etc.).

jason0x43 commented:
I haven’t forgotten this, I just haven’t had a chance to look at it yet. Thank you for contributing!

jason0x43 commented on custom_components/hubitat/event.py:
```diff
@@ -0,0 +1,100 @@
+        self._attr_event_types = self.get_event_types()
+
+    @property
+    def name(self) -> str:
```
Rather than overriding `name`, set `self._attr_name` in the constructor. That's what the other entity classes _should_ be doing, although not all of them are yet.

jason0x43 commented on custom_components/hubitat/event.py:
```diff
@@ -0,0 +1,100 @@
+        self.async_write_ha_state()
+
+    def get_event_types(self) -> list[str]:
+        event_attrs = filter(
```
This would be a bit simpler with list comprehensions, like

```py
event_attrs = [attr in self._device.attributes if attr in _TRIGGER_ATTR_MAP]
event_types = [_TRIGGER_ATTR_MAP[attr] for attr in event_attrs]
return event_types
```

or even

```py
return [_TRIGGER_ATTR_MAP[attr] for attr in self._device_attributes if attr in _TRIGGER_ATTR_MAP]
```

jason0x43 commented on custom_components/hubitat/event.py:
```diff
@@ -0,0 +1,100 @@
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+
+from .const import (
```
Only `_TRIGGER_ATTR_MAP` is actually needed in here; everything that's checking if an attribute is in `_TRIGGER_ATTRS` could instead be checking if it's in `_TRIGGER_ATTR_MAP`.

However, instead of importing these constants, it would be a bit cleaner to define a constant in this module that applies specifically to button event entities, like
```py
ATTR_EVENTS = {
  DeviceAttribute.DOUBLE_TAPPED: EventName:DOUBLE_TAPPED,
  DeviceAttribute.PUSHED: EventName.PUSHED,
  ...
}
```

This should include at least the 4 attributes and events used by Hubitat (double tapped, held, pushed, and released).

Then `handle_event` can do `if event.attribute not in ATTR_EVENTS` and `event_type = ATTR_EVENTS[event.attribute]`. It just makes the code in here a little easier to follow.

jason0x43 reviewed (changes requested):
This is great! I tested it locally, and everything seems to be working as expected. I just have a few minor requests to clean up the code a bit.

jason0x43 commented on custom_components/hubitat/const.py:
```diff
@@ -169,3 +170,9 @@ def __init__(
 }
+
+# Hubitat attributes that should be emitted as HA events
+_TRIGGER_ATTRS = tuple([v.attr for v in TRIGGER_CAPABILITIES.values()])
```
Hmmm...alternatively, it may be fine to move them from hub.py to const.py, but in that case we don't really need `_TRIGGER_ATTRS`, just `_TRIGGER_ATTR_MAP` (and without the leading underscore). We would also want to addd an entry to `TRIGGER_CAPABILITES` for `RELEASABLE_BUTTON`, so button released events would work.

kevdliu commented:
Thanks for the feedback! I'll update the pr when I get a chance.

kevdliu commented:
I think I addressed all the feedback, lmk if I missed any. The logic around event type handling does look a lot simpler now, thanks for the suggestions!

jason0x43 commented:
Thanks! This looks good. I'll try it out when I'm back with my Hubitat.
